src/ConstantBounds.cpp

```diff
@@ -0,0 +1,1141 @@
+                        // Lower bound: (x * c1) / c0 - (c1 + 1) <= (x / c0) * c1
+                        Expr offset = IntImm::make(op->type, factor - 1);
+                        return Sub::make(std::move(recurse_div), std::move(offset));
+                    } else {
+                        // Upper bound: (x / c0) * c1  <= (x * c1) / c0
+                        return recurse_div;
+                    }
+                }
+            }
+        }
+    }
+
+    // Base case. Stop pushing the multiplication inwards,
+    // but recurse on the inside if possible.
+    Expr expr_factor = IntImm::make(expr.type(), factor);
+    Expr recurse = handle_push_none(expr, direction);
+    return Mul::make(std::move(recurse), std::move(expr_factor));
+}
+
+Expr handle_push_none(const Expr &expr, Direction direction) {
+    debug(3) << "push_none(" << expr << ", " << direction << ")\n";
+    // Upper bound or lower bound without a factor or denominator
+    if (const Add *op = expr.as<Add>()) {
+        return handle_push_none(op->a, direction) + handle_push_none(op->b, direction);
+    } else if (const Sub *op = expr.as<Sub>()) {
+        return handle_push_none(op->a, direction) - handle_push_none(op->b, flip(direction));
+    } else if (const Min *op = expr.as<Min>()) {
+        return min(handle_push_none(op->a, direction), handle_push_none(op->b, direction));
+    } else if (const Max *op = expr.as<Max>()) {
+        return max(handle_push_none(op->a, direction), handle_push_none(op->b, direction));
+    } else if (const Select *op = expr.as<Select>()) {
+        const Expr true_value = handle_push_none(op->true_value, direction);
+        const Expr false_value = handle_push_none(op->false_value, direction);
+        return select(op->condition, true_value, false_value);
+    } else if (const Mul *op = expr.as<Mul>()) {
+        if (const IntImm *constant = op->b.as<IntImm>()) {
+            if (constant->value == 1) {
+                // strip off a mul-by-one.
+                return handle_push_none(op->a, direction);
+            } else {
+                return handle_push_mul(op->a, direction, constant->value);
+            }
+        } else if (const IntImm *constant = op->a.as<IntImm>()) {
+            if (constant->value == 1) {
+                // strip off a mul-by-one.
+                return handle_push_none(op->b, direction);
+            } else {
+                return handle_push_mul(op->b, direction, constant->value);
+            }
+        }
+    } else if (const Div *op = expr.as<Div>()) {
+        if (const IntImm *constant = op->b.as<IntImm>()) {
+            if (constant->value > 0) {
+                return handle_push_div(op->a, direction, constant->value);
+            } else {
+                Expr recurse = handle_push_none(op->a, flip(direction));
+                return Div::make(std::move(recurse), op->b);
+            }
+        }
+    }
+    return expr;
+}
+
+// Taken from Simplify_Let.cpp
```

Comment:
Move it into a common file, maybe ExprUsesVar.h/cpp

Reply:
done in bf131e4d0